### ISSUE TITLE
Concurrency race condition when using LocalStorage

### ISSUE THREAD

**Issue by mhesler** [simplabs/ember-simple-auth#254]
My Authenticator is resolving with a multi-value object that include the access_token, refresh_token, .expires, etc.

When a second window is opens the session restore process is initiated which ends up calling setup ->  store.replace ->  store.clear  just before the persist is called again.
In between the clear and persist, the other window storage event is triggered which ends up reading only a portion of the keys that are being persisted by the other window.  The sessionDataUpdated event is trigger which begins the same cycle in the other window.  This continues until one of them sees a version of the object without the authenticator property.  At that point it will clear the store.

Is there a reason we cannot store a JSON.stringify version of the object in a single key? With this approach there is no reason to do a clear first during the replace operation AND we will have a consistent version of the object to compare against and break the cycle.

**Comment by vdveer** [comment]
In some rare cases I still see a race condition between tokens. When an application forces a login by dropping a token, after authentication the 'storage' event triggers (session-stores, local-storage.js and validates this._lastData with the newly available storagedata. This event has old and new value available, but this can't be used in all browsers (IE has a bug there), so I am guessing this is why this._lastData has been choosen. But this var potentially has two different values in seperate tabs, in which case bad timing can cause an infinite (until it fails once) authenticator.restore() call with two switching data objects.
In my local project I have resolved this by putting the _lastData object also in a localstorage-record, which as far as I see now fixes this problem for me. I might be overlooking other usages of _lastData, so if you have so spare time, could you take a look at https://github.com/simplabs/ember-simple-auth/compare/master...vdveer:master ? Thanks @marcoow

**Comment by vdveer** [comment]
The race condition also triggered because I put a random value on the (non authenticated part) of the session before starting the authentication flow and remove it later on which also triggered a 'storage' event because now all session values use the same localstorage key.

**Comment by marcoow** [comment]
@vdveer: it's a bit hard to follow what exactly you're describing. Could you create a demo app that shows the behavior?
